File: src/data/blocks.ts
/**
 * The typed client for the document commands.
 *
 * Nodes cross the boundary as opaque JSON. The host stores them and does not
 * read them; what a block type means is the editor's business (ADR-003).
 */

import { invoke } from '@tauri-apps/api/core';

import type { Block, BlockChanges, DocNode } from '@/domain/document';
import type { Link } from '@/domain/page';

interface RawBlock {
  id: string;
  owner_kind: string;
  owner_id: string;
  type: string;
  position: string;
  content: DocNode;
}

function toBlock(raw: RawBlock): Block {
  return { id: raw.id, type: raw.type, position: raw.position, content: raw.content };
}

export async function listBlocks(ownerKind: string, ownerId: string): Promise<Block[]> {
  const raw = await invoke<RawBlock[]>('blocks_list', { ownerKind, ownerId });
  return raw.map(toBlock);
}

/**
 * Apply a change set, reindex, and record what the document points at, in one
 * host transaction.
 *
 * `plainText` is the flattened document and `links` the pages it links to.
 * Both travel with the change set so neither the search index nor the
 * backlinks can end up describing a document that was never saved.
 */
export async function applyBlocks(
  ownerKind: string,
  ownerId: string,
  changes: BlockChanges,
  plainText: string,
  links: readonly Link[] = [],
): Promise<Block[]> {
  const raw = await invoke<RawBlock[]>('blocks_apply', {
    ownerKind,
    ownerId,
    changes: {
      creates: changes.creates.map((create) => ({
        id: create.id,
        type: create.type,
        position: create.position,
        content: create.content,
      })),
      updates: changes.updates.map((update) => ({
        id: update.id,
        type: update.type,
        position: update.position,
        content: update.content,
      })),
      deletes: changes.deletes,
    },
    plainText,
    links: links.map((link) => ({ pageId: link.pageId, title: link.title })),
  });
  return raw.map(toBlock);
}

File: src-tauri/src/commands/blocks.rs
//! Document commands.

use tauri::State;

use crate::db::models::{Block, BlockChanges};
use crate::db::{blocks, Db};
use crate::error::Result;

#[tauri::command]
pub fn blocks_list(db: State<'_, Db>, owner_kind: String, owner_id: String) -> Result<Vec<Block>> {
    let conn = db.0.lock().expect("the database lock was poisoned");
    blocks::list_blocks(&conn, &owner_kind, &owner_id)
}

/// Apply a document change set and reindex, in one transaction.
///
/// `plain_text` is the flattened document, supplied by the caller: knowing that
/// a heading's text lives at `content[0].text` is the editor's business, not the
/// host's.
#[tauri::command]
pub fn blocks_apply(
    db: State<'_, Db>,
    owner_kind: String,
    owner_id: String,
    changes: BlockChanges,
    plain_text: String,
) -> Result<Vec<Block>> {
    let mut conn = db.0.lock().expect("the database lock was poisoned");
    blocks::apply_changes(&mut conn, &owner_kind, &owner_id, changes, &plain_text)
}
